File: LeetCode/139.Word_Break/139.Word_Break.py
#!/usr/bin/env python3


# tc: O(n * m) / n = len(s), m = len(wordDict)
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        # dp[n+1]
        # iterate word dict
        # for each word: dp[i] = dp[i-len(word)] if we can find s[i:i+len(word)] = curent word

        n = len(s)
        dp = [False] * (n + 1)
        dp[0] = True  # "" is True

        for i in range(1, n + 1):
            for w in wordDict:
                left = i - len(w)
                if left >= 0 and dp[left]:
                    if s[left:i] == w:
                        dp[i] = True
                if dp[i]:
                    break

        # print(dp)

        return dp[n]


# tc: O(n * n) / n = len(s)
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        # dp[n+1]
        # iterate word dict
        # for each word: dp[i] = dp[i-len(word)] if we can find s[i:i+len(word)] = curent word

        word_set = set(wordDict)
        n = len(s)
        dp = [False] * (n + 1)
        dp[0] = True  # "" is True

        for i in range(1, n + 1):
            for j in range(i):
                if dp[j] and s[j:i] in word_set:
                    dp[i] = True
                    break

        return dp[n]

File: LeetCode/139.Word_Break/139.Word_Break.swift
class Solution {
    func wordBreak(_ s: String, _ wordDict: [String]) -> Bool {
        // dp[i] represent s[0..i] can be segmented into dic or not
        let chs = Array(s)
        let n = chs.count
        let wordSet = Set(wordDict)

        var dp = Array(repeating: false, count: n + 1)

        // init dp, "" string
        dp[0] = true

        for right in 1 ... n {
            for word in wordDict {
                let len = word.count
                // cur word len > right
                if len > right { continue }

                let left = right - len
                if !dp[left] { continue }

                let subStr = String(chs[left ..< right])
                if word == subStr {
                    dp[right] = true
                    break
                }
            }
        }

        return dp[n]
    }
}
